src/tmpfiles/tmpfiles.c

```diff
@@ -2003,6 +2005,177 @@ static int glob_item_recursively(Item *i, fdaction_t action) {
+                        }
+                        return -errno;
+                }
+        }
+
+        /* Do not remove mount points. */
+        if (parent_st->st_dev != st.st_dev) {
+                log_error("mount point at \"%s/%s\"", strna(parent_name), name);
+                return -EBUSY;
+        }
+
+        if ((st.st_mode & S_IFMT) == S_IFDIR) {
+                _cleanup_close_ int child_fd = -1;
+
+                child_fd = openat(parent_fd, name, O_NOCTTY | O_CLOEXEC | O_DIRECTORY);
+                if (child_fd < 0)
+                        return log_error_errno(errno, "failed to open for rm-rf \"%s/%s\": %m", strna(parent_name), name);
+
+                r = rm_rf_children(TAKE_FD(child_fd), REMOVE_ROOT|REMOVE_SUBVOLUME|REMOVE_PHYSICAL, &st);
+                if (r < 0) {
+                        return log_error_errno(r, "failed to remove children \"%s/%s\": %m", strna(parent_name), name);
+                }
+
+                /* Ideally this would be unlinkat_harder but it isn't exported. */
+                r = unlinkat(parent_fd, name, AT_REMOVEDIR);
+        } else {
+                /* Ideally this would be unlinkat_harder but it isn't exported. */
+                r = unlinkat(parent_fd, name, 0);
+        }
+        if (r < 0) {
+                if (errno != ENOENT) {
+                        (void) log_error_errno(errno, "failed to unlink \"%s/%s\": %m", strna(parent_name), name);
+                }
+                return -errno;
+        }
+        return -ENOENT;
+}
+
+static void close_parent(int* parent_fd) {
+        if (*parent_fd != AT_FDCWD) {
+                safe_close(*parent_fd);
+        }
+}
+
+/* If child_mode is non-zero, rm_if_wrong_type_safe will be executed for the last path component. */
+static int mkdir_parents_rm_if_wrong_type(mode_t child_mode, const char *path) {
+        _cleanup_(close_parent) int parent_fd = -1;
+        _cleanup_close_ int next_fd = -1;
+        struct stat parent_st;
+        const char *s, *e;
+        int r;
+        size_t path_len;
+
+        assert(path);
+        assert((child_mode & ~S_IFMT) == 0);
+
+        path_len = strlen(path);
+
+        e = strrchr(path, '/');
+        if (!e)
+                return child_mode ? rm_if_wrong_type_safe(child_mode, AT_FDCWD, NULL, path, false) : 0;
+
+        /* Trailing path separators are only allowed if child_mode isn't set. */
+        if (((size_t)(e - path + 1) == path_len) && child_mode)
+                return -EINVAL;
+
+        /* Get the parent_fd and stat. */
+        parent_fd = AT_FDCWD;
+        if (*path == '/') {
+                parent_fd = open("/", O_NOCTTY | O_CLOEXEC | O_DIRECTORY);
+                if (parent_fd < 0)
+                        return log_error_errno(errno, "failed to open root: %m");
+                r = fstat(parent_fd, &parent_st);
+        } else {
+                r = stat("", &parent_st);
```

Comment:
hmm?

Reply:
Yeah, this isn't needed because of the AT_CWD.